### PR TITLE
Refactor Naming rules and add Variable/Function length rules

### PR DESCRIPTION
**Pull request by Mauin**
Introduces one MultiRule for all naming concerns in the `NamingRules` (that file is a bit huge now...)

Adds Min/Max name length rules for variables and functions (default values taken from PMD). 

Part of #302

### CODE REVIEW

**Comment by arturbosch** [comment]
Wow that looks really good! How about we introduce a new package `naming` inside style and split the whole file?

**Comment by schalkms** [comment]
It's a good idea to introduce a new naming package.

**Comment by Mauin** [comment]
Yep, I'll put them in their own `naming` package. WDYT about the defaults?

**Comment by arturbosch** [comment]
I think they are ok (pmd uses them). Personally I used longer names for functions and constants in detekt, so I would go for 25 chars :D. I don't know why we have that many detekt issues oO?

File: README.md [hunk 1]
```diff
@@ -425,13 +425,36 @@ style:
...
+  PackageNaming:
+    active: true
+    packagePattern: '^[a-z]+(\.[a-z][a-z0-9]*)*$'
+  ClassNaming:
     active: true
-    variablePattern: '^(_)?[a-z$][a-zA-Z$0-9]*$'
-    constantPattern: '^([A-Z_]*|serialVersionUID)$'
-    methodPattern: '^[a-z$][a-zA-Z$0-9]*$'
     classPattern: '[A-Z$][a-zA-Z$]*'
+  EnumEntryNaming:
```
Comment (vanniktech) [inline README.md]: I guess entry is only for each enum value so would it make sense to have a `EnumNaming` rule that only checks for the enum name?

File: README.md [hunk 2]
```diff
@@ -425,13 +425,36 @@ style:
...
+    active: true
     enumEntryPattern: '^[A-Z$][a-zA-Z_$]*$'
+  FunctionNaming :
+    active: true
+    functionPattern: '^[a-z$][a-zA-Z$0-9]*$'
+  FunctionMaxLength:
+    active: false
+    maximumFunctionNameLength: 17
```
Comment (vanniktech) [inline README.md]: seems quite low isn't it? I'd go for 32 probably

File: README.md [hunk 3]
```diff
@@ -425,13 +425,36 @@ style:
     active: true
   ModifierOrder:
     active: true
-  NamingConventionViolation:
+  PackageNaming:
+    active: true
+    packagePattern: '^[a-z]+(\.[a-z][a-z0-9]*)*$'
```
Comment (vanniktech) [inline README.md]: Seems inconsistent to allow digits in package names but not in class names.

File: README.md [hunk 4]
```diff
@@ -425,13 +425,36 @@ style:
...
-    variablePattern: '^(_)?[a-z$][a-zA-Z$0-9]*$'
-    constantPattern: '^([A-Z_]*|serialVersionUID)$'
-    methodPattern: '^[a-z$][a-zA-Z$0-9]*$'
     classPattern: '[A-Z$][a-zA-Z$]*'
+  EnumEntryNaming:
+    active: true
     enumEntryPattern: '^[A-Z$][a-zA-Z_$]*$'
+  FunctionNaming :
+    active: true
+    functionPattern: '^[a-z$][a-zA-Z$0-9]*$'
```
Comment (vanniktech) [inline README.md]: can't we use `\w` here?

File: README.md [hunk 5]
```diff
@@ -425,13 +425,36 @@ style:
...
+  FunctionMaxLength:
+    active: false
+    maximumFunctionNameLength: 17
+  FunctionMinLength:
+    active: false
+    minimumFunctionNameLength: 3
+  VariableNaming :
+    active: true
+    variablePattern: '^(_)?[a-z$][a-zA-Z$0-9]*$'
```
Comment (vanniktech) [inline README.md]: also here we could use `\w`

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+	private fun KtVariableDeclaration.isTopLevel(): Boolean = this is KtProperty && this.isTopLevel
+
+	companion object {
+		const val RULE_SUB_CONFIG = "NamingConventionViolation"
+	}
+}
+
+internal fun KtNamedDeclaration.identifierName(): String {
+	return nameIdentifier?.text ?: SpecialNames.NO_NAME_PROVIDED.asString()
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt]: inline?

Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt]: I meant to use the `() = nameIdentifier ..` syntax. Sorry for the confusion :D

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt [hunk 7]
```diff
@@ -168,6 +168,7 @@ object XmlEscape {
 /**
  * This class initializes the XML10_SYMBOLS structure.
  */
+@Suppress("ClassNaming")
```
Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt]: We can add here `all` as this was copied from github ^^

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	)
+
+	override fun postVisit(root: KtFile) {
+		super.postVisit(root)
+		rules.forEach { context.report(it.findings) }
+	}
+
+	override fun visitPackageDirective(directive: KtPackageDirective, data: Void?): Void? {
+		directive.reportFindings(context, packageNamingRule)
+		return null
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/naming/NamingRules.kt]: You can use the rule without the second parameter.

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt (hunk 7 above)
Comment (Mauin) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt]: Oh, the file is already annotated with `@file:Suppress("ALL")` and that doesn't seem to have any effect. Same if I replace `ClassNaming` with `all` here. Is suppressing everything still working?

Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt]: Yes suppress `ALL` on file level still works. There are 29 smells in this file. But `@Suppress(ClassNaming)` does work?

Comment (Mauin) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt]: Yeah, suppressing just that one rule works

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt [hunk 9]
```diff
@@ -168,7 +168,7 @@ object XmlEscape {
 /**
  * This class initializes the XML10_SYMBOLS structure.
  */
-@Suppress("ClassNaming")
+@Suppress("all")
```
Comment (arturbosch) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt]: I just checked that `all` is not supported, but `ALL`. We should support both.

Comment (Mauin) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/XmlEscape.kt]: Oh, okay. I'll change it to `all` and create another PR for `ALL` 👍

**Comment by Mauin** [comment]
The failures are coming from the `detektCheck` task which now reports NamingConvention issues.

**Comment by arturbosch** [comment]
Seems that the enum entry pattern is somehow broken?

**Comment by Mauin** [comment]
@arturbosch I believe CI is failing with the NamingViolation issues because:
- Travis is running `./gradlew detektCheck` on the plugin v1.0.0.RC2 (the current release).
- Since I renamed the naming rules, it cannot pick up the patterns from the `default-config` anymore
- The default pattern is different (only uppercase chars) from the `default-config` which allows lower case.

Now that the default config is changed, `detektCheck` reads the default naming Pattern which does not allow CamelCase but only UPPERCASE.

See: https://github.com/arturbosch/detekt/blob/master/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NamingConventionViolation.kt#L36

**Comment by Mauin** [comment]
The error is still because `detectCheck` uses the previous version of detekt. This will happen every time we change rule or rule parameter names.

**Comment by arturbosch** [comment]
Yeah, I understand. I will just merge this later. Hopefully I can release a RC4 soon with this and all the throw/catch multi rule changes which I'm working on right now.

**Comment by Mauin** [comment]
But I guess if we merge a change like this, every PR that happens before the next release will see the same issue in PR builds, won't they?

**Comment by arturbosch** [comment]
Yes, but there is nothing we can do right now. I have soon time in the train to work on detekt and hopefully be finished tomorrow in the evening. If you have have any suggestions for more naming changes now is the time :D

**Comment by Mauin** [comment]
I'm more thinking how we can prevent this from happening on CI. Running a different version of `detektCheck` maybe

**Comment by arturbosch** [comment]
We could change CI to use the build version via CLI and a subset of the rules via gradle detektCheck. I still want to run the Gradle plugin on the CI..

**Comment by schalkms** [comment]
@arturbosch
I think the exception multi rule is much needed.

**Comment by Mauin** [comment]
Oh, just saw these conflicts. I'll resolve them later today.

**Pull request merged**